[Clang][AArch64] Add customisable immediate range checking to NEON

SpencerAbson opened the pull request:
This patch moves NEON immediate argument specification and checking to the system currently shared by both SVE and SME. 

In its current form, the TableGen definition of a NEON intrinsic cannot control how its immediate arguments are range-checked, this information must be inferred from the name of the intrinsic by NeonEmitter, which also assumes that any NEON instruction will only ever receive a single immediate argument. For SVE/SME instrinsics, this information is more conveniently supplied in the TableGen definition.

As a result, for each immediate argument, NEON instructions must define
- The index of the immediate argument to be checked
- The type of immediate range check to be performed,
    (e.g., ImmCheckShiftRight)
- The index of the argument whose type defines the context
    of this immediate check (base type, vector size).
    - **Difference from SVE/SME** If this definition generates a polymorphic NEON builtin, the base type defined by this argument is overwritten by that of the type code supplied to the overloaded builtin call. This third argument is omitted in some cases due to this.

Here is an example for [`vfma_laneq`](https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[Neon]&q=vfma_laneq)
- The immediate is supplied in argument 3
- The immediate is used as an index into the lanes of argument 2
- So we must perform an immediate check on argument 3, based on the type information of argument 2.
- `ImmCheck<3, ImmCheckLaneIndex, 2>`

During this work, we discovered that the existing immediate range-checking system was largely untested, which made it difficult to make reliable progress. Missing tests have been added to verify this implementation against all intrinsics which take constrained immediate arguments. All test immediate range checking tests for NEON intrinsics are moved to a dedicated directory `clang/test/Sema/aarch64-neon-immediate-ranges/`.

SpencerAbson committed:
Rebase to resolve arm_neon.td conflict

SpencerAbson committed:
Updated/consistent vcmla codegen tests

SpencerAbson committed:
Run clang-format

Lukacma commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -810,7 +836,7 @@ bool SemaARM::CheckNeonBuiltinFunctionCall(const TargetInfo &TI,
   }
 
-  return SemaRef.BuiltinConstantArgRange(TheCall, i, l, u + l);
+  return ParseNeonImmChecks(TheCall, ImmChecks, TV);
```
I think using TV here is unnecessary. If we always specified the index of the base type for each intrinsics this could be removed and we could merge ParseNeon and ParseSVE

Lukacma commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,6 +369,63 @@ enum ArmSMEState : unsigned {
+    case ArmImmCheckType::ImmCheckLaneIndex:
+      hasError |= SemaRef.BuiltinConstantArgRange(
+          TheCall, ArgIdx, 0,
+          (64 << Type.isQuad()) / Type.getEltSizeInBits() - 1);
```
Shouldn't isQuad() be 0 for this case always ?

Lukacma commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,6 +369,63 @@ enum ArmSMEState : unsigned {
+      break;
+    case ArmImmCheckType::ImmCheckLaneQIndex: // force to use quad
+      hasError |= SemaRef.BuiltinConstantArgRange(
+          TheCall, ArgIdx, 0, (128 / Type.getEltSizeInBits()) - 1);
```
Once we stop relying on TV for base type we can move Type.getEltSizeInBits() into NeonEmitter and have the elements determined there. This will make it more similiar to SVE and enable merging of ParseNeon and ParseSVE

Lukacma reviewed (commented):
Thank you Spencer ! This patch is very useful as it will enable us to write more modular neon checks and unifies immediate checking with SVE.

SpencerAbson committed:
Move immediate check types to own generated file

Lukacma commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1327,49 +1392,68 @@ def SCALAR_RSHL: SInst<"vrshl", "11(S1)", "SlSUl">;
   // Signed/Unsigned Saturating Shift Right Narrow (Immediate)
-  def SCALAR_SQSHRN_N: SInst<"vqshrn_n", "(1<)1I", "SsSiSlSUsSUiSUl">;
+  def SCALAR_SQSHRN_N: SInst<"vqshrn_n", "(1<)1I", "SsSiSlSUsSUiSUl", 
+                            [ImmCheck<1, ImmCheckShiftRight, 0>]>;
```
Should be ImmCheckShiftRightNarrow

Lukacma commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1327,49 +1392,68 @@ def SCALAR_RSHL: SInst<"vrshl", "11(S1)", "SlSUl">;
   // Signed/Unsigned Saturating Rounded Shift Right Narrow (Immediate)
-  def SCALAR_SQRSHRN_N: SInst<"vqrshrn_n", "(1<)1I", "SsSiSlSUsSUiSUl">;
+  def SCALAR_SQRSHRN_N: SInst<"vqrshrn_n", "(1<)1I", "SsSiSlSUsSUiSUl", 
+                            [ImmCheck<1, ImmCheckShiftRight, 0>]>;
```
Should be ImmCheckShiftRightNarrow

Lukacma commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1327,49 +1392,68 @@ def SCALAR_RSHL: SInst<"vrshl", "11(S1)", "SlSUl">;
   // Signed Saturating Shift Right Unsigned Narrow (Immediate)
-  def SCALAR_SQSHRUN_N: SInst<"vqshrun_n", "(1<U)1I", "SsSiSl">;
+  def SCALAR_SQSHRUN_N: SInst<"vqshrun_n", "(1<U)1I", "SsSiSl", 
+                            [ImmCheck<1, ImmCheckShiftRight, 0>]>;
```
Should be ImmCheckShiftRightNarrow

Lukacma commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1327,49 +1392,68 @@ def SCALAR_RSHL: SInst<"vrshl", "11(S1)", "SlSUl">;
   // Signed Saturating Rounded Shift Right Unsigned Narrow (Immediate)
-  def SCALAR_SQRSHRUN_N: SInst<"vqrshrun_n", "(1<U)1I", "SsSiSl">;
+  def SCALAR_SQRSHRUN_N: SInst<"vqrshrun_n", "(1<U)1I", "SsSiSl", 
+                            [ImmCheck<1, ImmCheckShiftRight, 0>]>;
```
Should be ImmCheckShiftRightNarrow

Lukacma commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1959,9 +2064,12 @@ multiclass VCMLA_ROTS<string type, string lanety, string laneqty> {
       def : SOpInst<"vcmla" # ROT # "_laneq", "...QI", type,  Op<(call "vcmla" # ROT, $p0, $p1,
-              (bitcast $p0, (dup_typed lanety, (call "vget_lane", (bitcast laneqty, $p2), $p3))))>>;
-
+                (bitcast $p0, (dup_typed lanety, (call "vget_lane", (bitcast getlanety, $p2), $p3))))>>;
```
I am not sure I quite follow what is happening her. Could you elaborate on the purpose of getlanety ?

SpencerAbson replied:
The `vcmla` intrinsics are instantiated for each base type ([f16, f32, f64])  under this `VCMLA_ROTS` multiclass. For example:

`defm VCMLA_F32        : VCMLA_ROTS<"f", "uint64x1_t", "uint64x2_t">;`

The first argument is a vector with the number of lanes accessible by the `lane` argument in  [ vcmlaq_lane](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmlaq_lane)  and [vcmla_lane](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmla_lane) type inrinsics. The second argument is that for [vcmlaq_laneq](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmlaq_laneq) and [vcmla_laneq](https://developer.arm.com/architectures/instruction-sets/intrinsics/#q=vcmla_laneq) type intrinsics . This is done to give a vector of the correct number of elements to the call to `vget` in the instruction definition, so that the range of the immediate is correctly bounded (the range of this immediate is such that it can access some of the lower half of the vector only, which differs from traditional `_lane` type intrinsics where we access an arbitrary lane).

There is a problem with this approach for the `f16`/`h` base type; the range of the immediate in `vcmlaq_laneq_f16` is different to that for `vcmla_laneq_f16` (` 0<=lane<=3` and `0<=lane<=1` respectively). This is how the intrinsics are instantiated:

`defm VCMLA_FP16  : VCMLA_ROTS<"h", "uint32x2_t", "uint32x4_t">;`

The simplest fix I found for this case was to add the `getlanety` conditional which will select `uint32x2_t` for `vcmla_laneq_f16` and `uin32x4_t` for `vcmlaq_laneq_f16`, giving the correct range for the immediates.

Lukacma replied:
I did take a look into  [ACLE release](https://github.com/ARM-software/acle/releases/tag/r2024Q2) and I think the website you linked might have a mistake there. Looking at advsimd-2023Q2 I see `vcmla[_rot]_laneq_f16` defined like this :
<img width="725" alt="image" src="https://github.com/user-attachments/assets/2a446a5c-a3bf-40fc-9950-6caee95d81a2">

It also makes sense for all of them to have the same imm. range, as the range is only dependant on size and datatype of the 3 operand and not on the sizes and datatypes of the other operands.

SpencerAbson replied:
If we look at what this intrinsic compiles to, [FCMLA](https://developer.arm.com/documentation/ddi0596/2021-03/SIMD-FP-Instructions/FCMLA--by-element---Floating-point-Complex-Multiply-Accumulate--by-element--?lang=en), it looks like the reduced range (`0<=lane<=1`) is due to the fact that we must have `H=='0'` for this instruction to be defined when we use the `'01'` size specifier and `Q='0'` (what is required for this variant). This means that only bit-field `L` can be used to encode `<index>`.

Thanks to @rsandifo-arm for pointing this out.

Lukacma replied:
I am not sure I follow what you trying to say there. I think, it might be because your link to FCMLA is for the old spec and there may be some differences in instruction spec compared to the latest one. But if you look at the most recent spec for this [instruction](https://developer.arm.com/documentation/ddi0602/2024-06/SVE-Instructions/FCMLA--indexed---Floating-point-complex-multiply-add-by-indexed-values-with-rotate-), it says quite clearly that immediate range for half-precision variant is from 0 to 3

rsandifo-arm replied:
> [I]f you look at the most recent spec for this [instruction](https://developer.arm.com/documentation/ddi0602/2024-06/SVE-Instructions/FCMLA--indexed---Floating-point-complex-multiply-add-by-indexed-values-with-rotate-), it says quite clearly that immediate range for half-precision variant is from 0 to 3

That link is to the SVE instruction though, rather than the Advanced SIMD one.  The range for the 8H Advanced SIMD instruction is [0, 3], but the range for the 4H one is [0, 1].

Lukacma replied:
Apologies got confused there. Now I see what you are talking about. But to me it seems that whoever wrote this part of ACLE still wanted to allow lane to be from 0 to 3 and got around this limitation of the instruction by using DUP . So if we want to be consistent with ACLE we should still allow this wider range. Or do you think we should go and propose changing ACLE for this variant?

rsandifo-arm replied:
No, I agree that the ACLE expects an index of [2,3] to be emulated using DUP.

Lukacma reviewed (commented):
I think we should consider adding range tests for each intrinsics which specifies immediate range. This would help us detect errors in range checking specification, that are currently present.

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -333,6 +333,8 @@ class Intrinsic {
   /// The types of return value [0] and parameters [1..].
   std::vector<Type> Types;
+
+  SmallVector<std::tuple<int, int, int>, 2> ImmChecks;
```
Good point, I'll have SVEEmitter and NeonEmitter share ImmCheck.

SpencerAbson commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -1573,7 +1573,7 @@ void SVEEmitter::createTypeFlags(raw_ostream &OS) {
   OS << "#endif\n\n";
 
-  OS << "#ifdef LLVM_GET_SVE_IMMCHECKTYPES\n";
+  OS << "#ifdef LLVM_GET_ARM_INTRIN_IMMCHECKTYPES\n";
```
This was on my mind - done :)

SpencerAbson commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,6 +369,63 @@ enum ArmSMEState : unsigned {
   ArmZT0Mask = 0b11 << 2
 };
 
+bool SemaARM::ParseNeonImmChecks(
```
Well suggested, there is repetition here. Done.

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -447,19 +470,23 @@ def VST1_X3   : WInst<"vst1_x3", "v*(3!)",
 def VST1_X4_F16   : WInst<"vst1_x4", "v*(4!)", "hQh">;
-def VST1_LANE_F16 : WInst<"vst1_lane", "v*(.!)I", "hQh">;
+def VST1_LANE_F16 : WInst<"vst1_lane", "v*(.!)I", "hQh", 
+                          [ImmCheck<2, ImmCheck0_3, 1>]>;
```
I don't believe quad-variants currently exist for `vld1_lane` type intrinsics. But I will change this to the more general `ImmCheckLaneIndex`.

SpencerAbson replied:
Have changed them to use a more general `ImmCheckLaneIndex`. I overlooked `vst[x]q_lane_f16` in this approach, which was not flagged by any tests - these will be added too.

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -447,19 +470,23 @@ def VST1_X3   : WInst<"vst1_x3", "v*(3!)",
 def VLD1_X4_F16   : WInst<"vld1_x4", "4(c*!)", "hQh">;
-def VLD1_LANE_F16 : WInst<"vld1_lane", ".(c*!).I", "hQh">;
+def VLD1_LANE_F16 : WInst<"vld1_lane", ".(c*!).I", "hQh", 
+                          [ImmCheck<2, ImmCheck0_3, 1>]>;
```
I don't believe quad-variants currently exist for `vld1_lane` type intrinsics. But I will change this to the more general `ImmCheckLaneIndex`.

SpencerAbson replied:
Thank you!

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -473,44 +500,58 @@ def VLD3_DUP  : WInst<"vld3_dup", "3(c*!)",
+def VLD3_LANE_F16 : WInst<"vld3_lane", "3(c*!)3I", "hQh", 
+                          [ImmCheck<5, ImmCheck0_3, 1>]>;
+def VLD4_LANE_F16 : WInst<"vld4_lane", "4(c*!)4I", "hQh", 
+                          [ImmCheck<6, ImmCheck0_3, 1>]>;
```
I don't believe quad variants currently exist for these, but I have changed them to use a more general `ImmCheckLaneIndex`.

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -473,44 +500,58 @@ def VLD3_DUP  : WInst<"vld3_dup", "3(c*!)",
+                          [ImmCheck<3, ImmCheck0_3, 1>]>;
+def VST3_LANE_F16 : WInst<"vst3_lane", "v*(3!)I", "hQh", 
+                         [ImmCheck<4, ImmCheck0_3, 1>]>;
+def VST4_LANE_F16 : WInst<"vst4_lane", "v*(4!)I", "hQh", 
```
I don't believe quad variants currently exist for these, but I have changed it to use a more general `ImmCheckLaneIndex`.

SpencerAbson commented on clang/include/clang/Basic/arm_fp16.td:
```diff
@@ -76,7 +76,7 @@ let ArchGuard = "defined(__aarch64__) || defined(__arm64ec__)", TargetGuard = "f
   def SCALAR_FCVTPUH1 : SInst<"vcvtp_u32", "(1U>)1", "Sh">;
   def SCALAR_FCVTPUH2 : SInst<"vcvtp_u64", "(1U>>)1", "Sh">;
-  let isVCVT_N = 1 in {
+  let isVCVT_N = 1, ImmChecks = [ImmCheck<1, ImmCheck1_16>] in {
```
Thank you!

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -401,27 +405,45 @@ def VQRSHL : SInst<"vqrshl", "..S", "csilUcUsUiUlQcQsQiQlQUcQUsQUiQUl">;
+                      [ImmCheck<1, ImmCheckShiftRight>]>;
+def VRSHRN_N   : IInst<"vrshrn_n", "<QI", "silUsUiUl", 
+                      [ImmCheck<1, ImmCheckShiftRight>]>;
+def VQRSHRN_N  : SInst<"vqrshrn_n", "<QI", "silUsUiUl", 
```
These definitions generate polymorphic builtins, the type supplied used represent each overloaded variant is that of the **narrower return type**. This means that `ImmCheckShiftRightNarrow` would incorrectly apply an upper bound of `sizeinbits(elt)/4`, whereas `ImmCheckShiftRight` produces the correct behavior.
  
I will leave a comment about this caveat.

SpencerAbson committed:
Refactor immediate range checking change

For each immediate, a NEON instruction must define
  - The index of the immediate argument to be chcked

  - The type of immediate range check to be performed,
    (e.g., ImmCheckShiftRight)

  - The index of the argument whose type defines the context
    of this immediate check (base type, vector size).

	- Note that if this definition generates a polymorphic
	  NEON builtin, the base type defined by this argument is
	  overwritten by that of the type code supplied to the
	  overloaded builtin call. This third argument is omitted
	  in some cases due to this.

- Example
 - vfma_laneq
 (https://developer.arm.com/architectures/instruction-sets/intrinsics/#f:@navigationhierarchiessimdisa=[Neon]&q=vfma_laneq)
	- The immdiate is supplied in argument 3.
	- The immdiate is used as an index into the lanes of argument 2
	- So we must perform an immediate check on argument 3, based on
	  the number of lanes in argument 2.
	- ImmCheck<3, ImmCheckLaneIndex, 2>

- Changes
 - include/clang/Basic/arm_neon.td
	- Fix incorrect immediate range specifiers
	- Replace ImmCheckLaneQIndex occurences with
	  ImmCheckLaneIndex and (vector width is now inferred
	  from type context argument).
 - include/clang/Basic/arm_fp16.td
	- fix immediate range of fcvt instructions
 - include/clang/Basic/arm_immcheck_incl.td
	- Remove ImmCheckLaneQIndex
	- Add ImmCheckCvt
 - include/clang/Basic/TargetBuiltins.h
	- revert renaming of ImmCheckTypes

 - utils/Tablegen/NeonEMitter.cpp
	- Share ImmCheck class with SVEEmitter
	- Move parsing of ImmCheck list to constructor so that
	  TypeContextArg can be understood.
 - utils/TableGen/SVEEmitter.cpp
	- Reflect renaming of 'EltSizeArg' to 'TypeContextArg'.

 - Addition
  - llvm/include/llvm/TableGen/AArch64ImmCheck.h
	- move ImmCheck cpp class so that it can be shared
	  by NeonEmitter and SVEEmitter.

SpencerAbson committed:
Add NEON intrinsic immediate range tests

- Create ../clang/test/Sema/aarch64-neon-immediate-ranges/
	- including immediate range checking tests for all intrinsics
	  that take an immediate argument and aren't tested elsewhere.
	- A note has been made where tests have been excluded to avoid
	  duplication, and where these existing tests have been found.
	- A note has been made about the exclusion of widening left-shifts
	  from these tests.
	- These tests were created with the help of this tool (written by
	  the author of this commit), https://github.com/SpencerAbson/imm-test-gen
		- Although they are generated on a per-file basis, care has been
		  taken to ensure that the tests are correct and no test
		  duplication is introduced.

Lukacma commented on clang/test/Sema/aarch64-neon-vcmla-ranges.c:
```diff
@@ -0,0 +1,202 @@
+// RUN: %clang_cc1 -triple aarch64-linux-gnu -target-feature +neon -target-feature +v8.3a -ffreestanding -fsyntax-only -verify %s
```
Why isn't this test under this path ? clang/test/Sema/aarch64-neon-immediate-ranges

SpencerAbson replied:
Moved :+1:

Lukacma commented on clang/test/CodeGen/aarch64-neon-vcmla.c:
```diff
@@ -1,444 +1,596 @@
-// RUN:        -target-feature +fullfp16 \
-// RUN:        -disable-O0-optnone -emit-llvm -o - %s | opt -S -O1 | FileCheck %s
+// NOTE: Assertions have been autogenerated by utils/update_cc_test_checks.py UTC_ARGS: --version 5
+// RUN: %clang_cc1 -triple arm64-linux-gnu -target-feature +neon  \
```
Why was the run line changed here ?

SpencerAbson replied:
Just an artifact of recreating that test file - though I have reverted this file back to it's previous state following our resolution of `vcmla_laneq_f16`, which was the reason this was originally changed at all.

SpencerAbson commented on clang/include/clang/Basic/arm_fp16.td:
```diff
@@ -76,17 +76,22 @@ let ArchGuard = "defined(__aarch64__) || defined(__arm64ec__)", TargetGuard = "f
+    def SCALAR_FCVTZUH1O: SInst<"vcvt_n_u32", "(1U>)1I", "Sh",
+                                [ImmCheck<1, ImmCheckCvt, 0>]>;
+    def SCALAR_FCVTZUH2O: SInst<"vcvt_n_u64", "(1U>>)1I", "Sh",
+                                [ImmCheck<1, ImmCheckCvt, 0>]>;
```
Good point, that flag was only ever used to inform the previous range-checking code. I'll remove it.

SpencerAbson commented on clang/include/clang/Basic/arm_immcheck_incl.td:
```diff
@@ -0,0 +1,39 @@
+def ImmCheckLaneIndex           : ImmCheckType<7>;  // 0..(128/(1*sizeinbits(elt)) - 1)
+def ImmCheckCvt                 : ImmCheckType<8>;  // 1..sizeinbits(elt) (same as ShiftRight)
+def ImmCheckLaneIndexCompRotate : ImmCheckType<9>;  // 0..(128/(2*sizeinbits(elt)) - 1)
+def ImmCheckLaneIndexDot        : ImmCheckType<10>; // 0..(128/(4*sizeinbits(elt)) - 1)
```
True, I think `sizeinbits(vec)` is the most consistent to what we are using already.

SpencerAbson commented on clang/include/clang/Basic/arm_immcheck_incl.td:
```diff
@@ -0,0 +1,39 @@
+def ImmCheck0_63                : ImmCheckType<26>; // 0..63
+
+class ImmCheck<int arg, ImmCheckType kind, int typeArg = -1> {
+  int Arg = arg;
```
:+1:  I'll also do this for `typeArg`.

SpencerAbson commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,142 +368,183 @@ enum ArmSMEState : unsigned {
+      return true;
+    break;
+  case ImmCheckType::ImmCheckCvt:
+  case ImmCheckType::ImmCheckShiftRight:
```
Ok sounds good, they both do `1..sizeinbits(elt)` so I could replace them both with `ImmCheck1_EltSize`.

Lukacma replied:
Maybe. You could rename them if you can find better naming for them, but I am fine with keeping them the same as well. We have other Immediate ranges which have cryptic names, so it is not a big issue I think.

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -1952,6 +1974,7 @@ void NeonEmitter::createIntrinsic(Record *R,
   bool BigEndianSafe  = R->getValueAsBit("BigEndianSafe");
   std::string ArchGuard = std::string(R->getValueAsString("ArchGuard"));
   std::string TargetGuard = std::string(R->getValueAsString("TargetGuard"));
+
```
thanks

SpencerAbson commented on clang/utils/TableGen/TableGen.cpp:
```diff
@@ -234,6 +235,9 @@ cl::opt<ActionType> Action(
         clEnumValN(GenArmNeonTest, "gen-arm-neon-test",
                    "Generate ARM NEON tests for clang"),
+        clEnumValN(GenArmImmCheckTypes, "gen-arm-immcheck-types",
+                   "Generate arm_immchecktypes.h (immediate range check types)"
```
thanks!

SpencerAbson commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -388,6 +372,9 @@ class SVEEmitter {
   /// Emit all the range checks for the immediates.
   void createRangeChecks(raw_ostream &o);
 
+  // Emit all the ImmCheckTypes to arm_immcheck_types.h
```
thanks!

SpencerAbson committed:
Remove redundant immrange flags from NEON instructions

- Changes
	- clang/include/clang/Basic/arm_immcheck_incl.td
	- clang/include/clang/Basic/arm_neon.td
		- Restore immediate range of vcmla{_ROT}_laneq_f16 intrinsics to [0..3]
	- clang/include/clang/Basic/arm_fp16.td
	- clang/include/clang/Basic/arm_neon_incl.td
	- clang/utils/TableGen/NeonEmitter.cpp
		- The flags 'isVXAR', 'isVCVT_N', and 'isLaneQ' erve no justifiable purpose
	          following the work of this patch, and so are removed from NEON instructions.

	- clang/include/clang/Basic/arm_immcheck_incl.td
	- clang/utils/TableGen/SveEmitter.cpp
		- The fields of the ImmCheck tablegen class are also renamed to better explain
		  their purposes.
		- The description of ImmCheckTypes is updated to reflect that we are now sharing
		  them with NEON.
	- clang/test/CodeGen/aarch64-neon-vcmla.c
		- Revert file main following resolution of vcmla{_ROT}_laneq_f16 range

SpencerAbson committed:
Update and move vcmla immediate range tests

	- update tests for vcmla{_ROT}_laneq_f16 after immedaite range resolution
	- move test file to clang/test/Sema/aarc64-neon-immedaite-ranges
	- remove trailing whitespace from test lines

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -369,12 +372,13 @@ class Intrinsic {
-        BigEndianSafe(BigEndianSafe), PolymorphicKeyType(0), NeededEarly(false),
-        UseMacro(false), BaseType(OutTS, "."), InBaseType(InTS, "."),
-        Emitter(Emitter) {
+        ArchGuard(ArchGuard.str()), TargetGuard(TargetGuard.str()),
```
clang-format I'm afraid.

SpencerAbson committed:
[Fixup] tablegen command descr and SemaArm

SpencerAbson committed:
Add check to ensure builtin range is not redefined

- Changes
	- llvm/inlcude/llvm/TableGen/AArch64ImmCheck.h
		- Add TypeArgIdx field for ImmCheck (third argument in tablegen)
		- introduce equality operator for ImmCheck
		- rename 'Arg' to 'ImmArgIdx' for consistency with the tablegen class.
	- clang/utils/TableGen/NeonEmitter.cpp
		- Add assertion to ensure that a neon builtin's range is not
		  redefined. This problem could arise when two intrinsics share the same builtin
		  but define different immediate checks in tablgen (range checking is done via
		  the builtin rather than the intrinsic).
	- clang/include/clang/Basic/arm_neon.td
		- ensure consistent TypeContextArg is used for vsri/vsli intrinsics.

SpencerAbson committed:
[NFC] fix format

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,84 +2159,44 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
-    if (Emitted.find(Def->getMangledName()) != Emitted.end())
+  for (auto &Def : Defs) {
+    if (Emitted.find(Def->getMangledName()) != Emitted.end() ||
+        !Def->hasImmediate())
```
Added now

SpencerAbson committed:
Fix builtin range-check compatabilty

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2169,35 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
-  std::set<std::string> Emitted;
+void NeonEmitter::genIntrinsicRangeCheckCode(
+    raw_ostream &OS, SmallVectorImpl<Intrinsic *> &Defs) {
+  std::map<std::string, ArrayRef<ImmCheck>> Emitted;
```
Good point, done.

SpencerAbson committed:
[Fixup] Improve comments and remove trailing whitespace

SpencerAbson commented on clang/include/clang/Basic/arm_neon.td:
```diff
@@ -1957,15 +1945,13 @@ multiclass VCMLA_ROTS<string type, string lanety, string laneqty> {
+    // vcmlaq{ROT}_laneq
+    def : SOpInst<"vcmla" # ROT # "_laneq", "...QI", "Q" # type, Op<(call "vcmla" # ROT, $p0, $p1,
+            (bitcast $p0, (dup_typed laneqty , (call "vget_lane", (bitcast laneqty, $p2), $p3))))>>;
 
```
Removed :+1:

Lukacma commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
-                                        SmallVectorImpl<Intrinsic *> &Defs) {
-  OS << "#ifdef GET_NEON_IMMEDIATE_CHECK\n";
+inline bool
+NeonEmitter::areCompatableRangeChecks(const ArrayRef<ImmCheck> ChecksA,
```
typo. Should be areCompatibleRangeChecks . Though maybe areRangeChecksCompatible would be better name.

SpencerAbson replied:
Done

Lukacma commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
+NeonEmitter::areCompatableRangeChecks(const ArrayRef<ImmCheck> ChecksA,
+                                      const ArrayRef<ImmCheck> ChecksB) {
+  // If multiple intrinsics map to the same builtin, we must ensure that the
+  // intended range checks performed in SemaArm.cpp do not contradict eachother,
```
eachother -> each other

SpencerAbson committed:
Unify NEON immediate range checking tests

- Changes
	- move existing tests to new dedicated directory
	- Fix NeonEmitter typos
	- Change vector store test prefix to 'test_vector_store'
	- Remove trailing whitespace in arm_neon.td

momchil-velikov commented on clang/include/clang/Sema/SemaARM.h:
```diff
@@ -37,15 +37,20 @@ class SemaARM : public SemaBase {
                                     CallExpr *TheCall);
+  bool
+  ParseNeonImmChecks(CallExpr *TheCall,
+                     SmallVector<std::tuple<int, int, int, int>, 2> &ImmChecks,
```
Should be `SmallVectorImpl`.

SpencerAbson replied:
Done.

momchil-velikov commented on clang/include/clang/Sema/SemaARM.h:
```diff
@@ -37,15 +37,20 @@ class SemaARM : public SemaBase {
                                     CallExpr *TheCall);
+  bool
+  ParseNeonImmChecks(CallExpr *TheCall,
+                     SmallVector<std::tuple<int, int, int, int>, 2> &ImmChecks,
```
Perhaps `std::tuple<int, int, int ...` warrants its own type name, no strong opinion either way.

SpencerAbson replied:
I would have done this if the type name could be shared SVE/SME and NEON, but this isn't the case as we omit the last value from the tuple in SME/SVE as it is always 128.

momchil-velikov commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,142 +369,183 @@ enum ArmSMEState : unsigned {
-      break;
-    }
+  for (const auto &I : ImmChecks) {
+    std::tie(ArgIdx, CheckTy, ElementSizeInBits) = I;
+    HasError |=
```
LIkewise, can use structured binding declaration.

SpencerAbson replied:
Done.

momchil-velikov commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -403,6 +408,38 @@ class Intrinsic {
+        VecSizeInBits = ContextType.getSizeInBits();
+      }
+
+      ImmChecks.emplace_back(
```
Just `.emplace_back(ArgIdx, Kind, ...`

SpencerAbson replied:
Done.

momchil-velikov commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
+  // are not and so must be the same.
+  bool compat =
+      std::equal(ChecksA.begin(), ChecksA.end(), ChecksB.begin(), ChecksB.end(),
+                 [](const auto A, const auto B) {
```
`const auto &A` (which is ofc moot if you decide to add `operator==`)

momchil-velikov commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -1210,18 +1196,19 @@ void SVEEmitter::createIntrinsic(
       unsigned NumVectors;
-      std::tie(Mod, NumVectors) = getProtoModifier(Proto, EltSizeArg + 1);
-      if (EltSizeArg >= 0)
+      std::tie(Mod, NumVectors) = getProtoModifier(Proto, EltSizeArgIdx + 1);
```
Could be `auto [Mod, NumVectors] = ...`

SpencerAbson replied:
Thanks, done.

momchil-velikov commented on clang/utils/TableGen/SveEmitter.cpp:
```diff
@@ -1541,8 +1528,9 @@ void SVEEmitter::createRangeChecks(raw_ostream &OS) {
     for (auto &Check : Def->getImmChecks())
-      OS << "ImmChecks.push_back(std::make_tuple(" << Check.getArg() << ", "
-         << Check.getKind() << ", " << Check.getElementSizeInBits() << "));\n";
+      OS << "ImmChecks.push_back(std::make_tuple(" << Check.getImmArgIdx()
```
Likewise, `.emplace_back(...`

SpencerAbson replied:
Thanks, done.

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
+  // the same. The element types may differ as they will be resolved
+  // per-intrinsic as overloaded types by SemaArm.cpp, though the vector sizes
+  // are not and so must be the same.
+  bool compat =
```
We actually had this approach initially, the problem with it is that this issue only prevalent for NEON intrinsics yet the class is also used for SVE/SME. Defining **equality** for SVE/SME and NEON based on a definition of **compatibility** that only applies to NEON intrinsics didn't seem right at the time.

SpencerAbson committed:
[Fixup] Improve semantics and remove unnecessary code

Changes
	- clang/lib/Sema/SemaARM.cpp clang/include/clang/Sema/SemaARM.h
		- Rename immediate checking functions from 'Parse' to 'Perform'
		- Change CheckImmediateInSet to test against a set of values, rather than
		  an expression.
		- Use SmallVectorImpl in SemaARM.h.
	- clang/utils/TableGen/NeonEmitter.cpp clang/utils/TableGen/SveEmitter.cpp
		- Simplify emitted expressions and unpacking values

SpencerAbson commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -403,142 +369,183 @@ enum ArmSMEState : unsigned {
+
+  // Function that checks whether the operand (ArgIdx) is an immediate
+  // that is one of the predefined values.
+  auto CheckImmediateInSet = [&](OptionSetCheckFnTy CheckImm,
```
Good point, done.

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
+    // Emit builtin's range checks
+    OS << "case NEON::BI__builtin_neon_" << Def->getMangledName() << ":\n";
+    for (const auto &Check : Checks) {
+      OS << " ImmChecks.push_back(std::make_tuple(" << Check.getImmArgIdx()
```
Thanks, done.

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
-  std::set<std::string> Emitted;
+void NeonEmitter::genIntrinsicRangeCheckCode(
+    raw_ostream &OS, SmallVectorImpl<Intrinsic *> &Defs) {
+  std::unordered_map<std::string, ArrayRef<ImmCheck>> Emitted;
```
Good point, unfortunately there is no standard implementation of `DenseMapInfo<std::string>` (required to have `std::string` as a key for `DenseMap`) - I'm not sure that this use would justify defining one here?

SpencerAbson commented on clang/utils/TableGen/NeonEmitter.cpp:
```diff
@@ -2142,85 +2178,58 @@ void NeonEmitter::genOverloadTypeCheckCode(raw_ostream &OS,
-    if (!UpperBound.empty())
-      OS << " u = " << UpperBound << ";";
-    OS << " break;\n";
+    const auto it = Emitted.find(Def->getMangledName());
```
I agree that this would be cleaner, but it's convenient to have the iterator returned by `find` and do the lookup just once in the case that this succeeds.

SpencerAbson committed:
Rebase and provide ImmChecks for NEON LUTI intrinsics
	- Add immediate range checking to new NEON LUTI intrinsics
	  (https://github.com/ARM-software/acle/commit/e938350b3034512912eb2655402324750cb8c1ef)
	- Add tests for this under clang/Sema/aarch64-neon-immediate-ranges/luti.c
	- Update CodeGen tests for NEON LUTI intrinsics to use legal indices
	- Use std::string::contains instead of ends_with to detect laneq in NEON operation DAG call
          generation code (this has no effect on these intrinsics (or any existing intrinsics) as they
          do not have operation DAGs, but the '_x2' naming convention prompted this action.

CarolineConcatto commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -404,142 +370,174 @@ enum ArmSMEState : unsigned {
+bool SemaARM::PerformNeonImmChecks(
+    CallExpr *TheCall,
+    SmallVectorImpl<std::tuple<int, int, int, int>> &ImmChecks,
+    int OverloadType = -1) {
```
Can you change this to be
bool SemaARM::PerformNeonImmChecks(
    CallExpr *TheCall,
    SmallVectorImpl<std::tuple<int, int, int, int>> &ImmChecks,
    int OverloadType)
and then in SemaARM.h
 bool PerformNeonImmChecks(
      CallExpr *TheCall,
      SmallVectorImpl<std::tuple<int, int, int, int>> &ImmChecks,
      int OverloadType=-1);

CarolineConcatto reviewed (approved):
Thank you Spencer for the patch.
I left some nits, but you can fix it before you merge the patch.

SpencerAbson committed:
[NFC] Remove trailing whitespace and move default overload value

SpencerAbson commented:
Thank you, @nico . The '-WError' buildbot failiures have been fixed since https://github.com/llvm/llvm-project/commit/383057e5a0b53a79926dbdbae4cfcd39bfb44645.

jroelofs commented on clang/lib/Sema/SemaARM.cpp:
```diff
@@ -404,142 +370,174 @@ enum ArmSMEState : unsigned {
+        SemaRef.BuiltinConstantArgMultiple(TheCall, ArgIdx, 2))
+      return true;
+    break;
+  default:
```
This causes a `-Wcovered-switch-default`.

jroelofs replied:
oh, never mind, nico fixed this already.

SpencerAbson replied:
I removed this here https://github.com/llvm/llvm-project/commit/383057e5a0b53a79926dbdbae4cfcd39bfb44645.

jroelofs commented on clang/include/clang/Basic/TargetBuiltins.h:
```diff
@@ -216,6 +216,36 @@ namespace clang {
+        return 64;
+      case Poly128:
+        return 128;
+      default:
```
This causes a `-Wcovered-switch-default`

jroelofs replied:
oh, never mind, nico fixed this already.
